sdk/python/lib/pulumi/x/automation/server.py

```diff
@@ -0,0 +1,80 @@
+# Copyright 2016-2020, Pulumi Corporation.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Optional
+
+from .workspace import PulumiFn
+from ...errors import RunError
+from ...log import *
+from ...runtime.proto import language_pb2, plugin_pb2, LanguageRuntimeServicer
+from ...runtime import run_in_stack, reset_options, set_all_config
+
+
+class LanguageServer(LanguageRuntimeServicer):
+    program: PulumiFn
+
+    pulumi_exit_code: Optional[int]
+    running: bool
+
+    def __init__(self, program: PulumiFn) -> None:
+        self.program = program  # type: ignore
+        self.running = False
+        self.pulumi_exit_code = None
+
+    def get_exit_error(self, preview: bool):
+        if self.pulumi_exit_code == 0:
+            return "pulumi exited prematurely"
+        op = "preview" if preview else "update"
+        return f"{op} failed with code {self.pulumi_exit_code}"
+
+    def on_pulumi_exit(self, code: int, preview: bool):
+        self.pulumi_exit_code = code
+
+        # Reset globals
+        reset_options()
+        if not self.running:
+            raise RunError(self.get_exit_error(preview))
+
+    def GetRequiredPlugins(self, request, context):
+        return language_pb2.GetRequiredPluginsResponse()
+
+    def Run(self, request, context):
```

Comment:
I have some questions about the error handling strategy here. I'm not overly familiar with the python runtime, so please excuse any silly questions I may be asking. I think it's worth talking through given some of the issues that have popped up with node recently.

In typescript, we add a set of uncaught exception handlers in addition to an outer try/catch block: 

https://github.com/pulumi/pulumi/blob/39297cb9ddb4e2499db72dc64175b6352e9ab7a8/sdk/nodejs/x/automation/server.ts#L96-L117

It doesn't seem like we have anything equivalent. It appears on the surface that an unhandled exception in user code could potentially bubble up unhandled, and short circuit our calls to `on_exit`. `run_pulumi_func` does have a try/catch block, but it can rethrow: 

https://github.com/pulumi/pulumi/blob/f36431106443fc431076b43df68b03538eae0855/sdk/python/lib/pulumi/runtime/stack.py#L74

I guess this isn't so much a question, as it is a desire to better understand the error handling strategy. Do we know what happens if we write a `program` that throws an exception or fails for some reason? 

If there's uncertainty here, we may be able to work through some of this together. I need to dive into the typescript error handling approach tomorrow anyways. I don't have a great deal of familiarity with gRPC so I'm not sure if the original implementations really are handling errors in the server appropriately. I did find that guide that has some code snippets for multi-language server error handling, but I'm not yet sure if these are applicable to our use case. https://avi.im/grpc-errors/#python

Reply:
Yeah, I kinda rushed through this bit - will spend some time looking into error handling and add some bad tests.